### PR TITLE
Compress crash bug refactor

### PR DESCRIPTION
**Pull request by aclamk**
This is a refactor of pull request #15335 .
Logic simplified, now rgw_compression produces chunks of size exactly `rgw_max_chunk_size`.

### CODE REVIEW

**Comment by tchaikov** [comment]
retest this please.

File: src/test/rgw/test_rgw_compression.cc [hunk 1]
```diff
@@ -48,3 +115,87 @@ TEST(Decompress, FixupRangePartial)
   ASSERT_EQ(range_t(12, 24), fixup_range(&decompress, 16, 999));
   ASSERT_EQ(range_t(18, 24), fixup_range(&decompress, 998, 999));
 }
+
+TEST(Compress, LimitedChunkSize)
+{
+  CompressorRef plugin;
+  plugin = Compressor::create(g_ceph_context, Compressor::COMP_ALG_ZLIB);
+  ASSERT_NE(plugin.get(), nullptr);
```
Comment (cbodley) [inline src/test/rgw/test_rgw_compression.cc]: i think this depends on the test being able to locate a valid ceph.conf. without one, this assertion fails

[test_async_compressor.cc](https://github.com/ceph/ceph/blob/master/src/test/common/test_async_compressor.cc), for example, has to set `plugin_dir` explicitly in order to run as part of `make check`. could you try that here?

Comment (aclamk) [inline src/test/rgw/test_rgw_compression.cc]: @cbodley 
Actually, logic for setting plugin_dir is already there. It comes from src/test/unit.cc.
https://github.com/ceph/ceph/blob/master/src/test/unit.cc#L44-L48
There is difference in linking between test_rgw_compression.cc and test_async_compressor.cc

Comment (aclamk) [inline src/test/rgw/test_rgw_compression.cc]: @cbodley 
What I can do, I can add logic to use ".libs" if CEPH_LIB is not set. Should I?

Comment (cbodley) [inline src/test/rgw/test_rgw_compression.cc]: oh cool, i didn't see that in `src/test/unit.cc`. that should be good enough

**Comment by cbodley** [comment]
passed rgw suite: http://pulpito.ceph.com/cbodley-2017-06-19_13:44:54-rgw-wip-cbodley-testing-distro-basic-mira/

**Comment by cbodley** [comment]
jenkins test this

**Comment by cbodley** [comment]
jenkins test this please

**Comment by cbodley** [comment]
jenkins test this please

**Comment by aclamk** [comment]
jenkins test this please

**Comment by cbodley** [comment]
jenkins test this please

**Comment by cbodley** [comment]
@aclamk needs rebase again :(

**Pull request merged**